proposals/0014-wave-size-range.md

```diff
@@ -0,0 +1,283 @@
+This increases the size of the shipped product and potentially slows down
+ runtime shader loading.
+
+Ranges of wave sizes are already present in areas of the D3D API.
+The query to determine what wave sizes are supported takes a range and similar
+ information is transmitted from the shader to the runtime, though not the
+ driver.
+
+## Proposed solution
+
+Modifying the parameters of the `WaveSize` attribute that was introduced by
+ Shader Model 6.6 to take additional parameters will allow the shader author to
+ specify fully what the shader supports and what it prefers.
+
+By adding an optional second `WaveSize` parameter, shader authors can provide
+ two values that represent the range of wave sizes that the shader can support.
+This allows the same shader to be used in the event of the availability of a
+ wave size that the shader has optimized for specifically as well as other sizes
+ that are supported, but might not be so optimal.
+
+Some platforms that support a range of wave sizes might overlap with the range
+ specified as supported by the shader in more than one value.
+In these cases, the runtime has a choice of wave sizes that it probably doesn't
+ have the information needed to choose the best wave size.
+To give the runtime the needed information, it is useful to specify the
+ preferred wave size in addition to the full range of acceptable values.
+By adding an optional third `WaveSize` parameter, shader authors can specify the
+ optimal wave size for a given shader in addition to the range of acceptable
+ sizes.
+
+## Detailed Design
+
+### HLSL additions
+
+The existing `WaveSize` attribute gains a second overload:
+
+```HLSL
+[WaveSize(<minWaveSize>, <maxWaveSize>, [prefWaveSize])]
+void main() ...
+```
+
+Where `minWaveSize` is the minimum wave size supported by the shader
+ representing the beginning of the allowed range,
+ `maxWaveSize` is the maximum wave size supported by the shader
+ representing the end of the allowed range,
+ and `prefWaveSize` is the optional preferred wave size representing the size
+ expected to be the most optimal for this shader.
+
+### DXIL Additions
+
+The existing `WaveSize` value is stored as a metadata scalar constant.
+To store the additional values, an additional metadata tag indicating a tuple of
+ three values representing the parameters to `WaveSize` given by the user.
+In the case where no preferred wave size is specified, the value zero will
+ indicate that nothing was specified.
+
+### SPIR-V Additions
+
+To make use of this attribute, SPIRV would have to represent it in new SPIRV
+ operations that drivers can read to determine what the values are an how to
+ use to them to choose the best wave size for a given shader.
+
+### Diagnostic Changes
+
+#### New Errors
+
+These are where new or slightly altered errors are produced:
+
+* If any of the parameters of `WaveSize` are not literal power of two integers
+  between 4 and 128 inclusive.
```

Comment:
Can we allow for explicit `0` for preferred to indicate no preference?  Isn't that what we encode when not included?

Reply:
I'd almost rather we make `0` an error. I can imagine in the future we might fix the bugs in DXC to allow `constexpr` values to be used for these arguments and it might be best to error on all values that are out of the [min, max] range so that logic errors in compile-time expressions get surfaced.

Reply:
I've filed #157 to follow up on this issue later.